src/assimp.cc

```diff
@@ -0,0 +1,419 @@
+#include "assimp.hh"
+#include "log.hh"
+#include "stb_image.h"
+#include <assimp/Importer.hpp>
+#include <assimp/scene.h>
+#include <assimp/postprocess.h>
+#include <numeric>
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace
+{
+using namespace tr;
+
+// FIXME: I've been copied from gltf.cc! Make this a utility function
+template<typename T>
+auto add_unique_named(std::string& name, T& map) -> decltype(map[name])&
+{
+    std::string candidate_name = name;
+    int count = 0;
+    while(map.count(candidate_name) != 0)
+    {
+        candidate_name = name + std::to_string(count++);
+    }
+    name = candidate_name;
+    return map[name];
+}
+
+template<typename T>
+std::string gen_free_name(std::string& name, T& map)
+{
+    std::string candidate_name = name;
+    int count = 0;
+    while(map.count(candidate_name) != 0)
+    {
+        candidate_name = name + std::to_string(count++);
+    }
+    return candidate_name;
+}
+
+std::vector<mesh::vertex> read_vertices(aiMesh* ai_mesh)
+{
+    std::vector<mesh::vertex> mesh_vert;
+    mesh_vert.reserve(ai_mesh->mNumVertices);
+
+    for(unsigned int j = 0; j < ai_mesh->mNumVertices; j++)
+    {
+        mesh::vertex v;
+        v.pos = vec3(
+            ai_mesh->mVertices[j].x,
+            ai_mesh->mVertices[j].y,
+            ai_mesh->mVertices[j].z
+        );
+
+        if(ai_mesh->HasNormals())
+            v.normal = vec3(
+                ai_mesh->mNormals[j].x,
+                ai_mesh->mNormals[j].y,
+                ai_mesh->mNormals[j].z
+            );
+        
+        if(ai_mesh->HasTangentsAndBitangents())
+            v.tangent = vec4(
+                ai_mesh->mTangents[j].x,
+                ai_mesh->mTangents[j].y,
+                ai_mesh->mTangents[j].z,
+                // Not sure if this is correct
+                1.0f
```

Comment:
It's indeed not, and the correct value can be calculated from the mBitangents member. See https://github.com/vga-group/tauray/blob/ad3773041877e16be9d8157b1ad973d50e1fc98d/shader/forward.vert#L32 for how it's used. 
I think the correct value here would be `sign(dot(ai_mesh->mBitangents[j], cross(ai_mesh->mNormals[j], ai_mesh->mTangents[j])))`.